src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,238 @@
+
+            JWTSigningKeyInfo keyInfo = null;
+
+            while (true) {
+                List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+                // Loop through the keys and find the first one for the algorithm, if the list is empty a new key will be created after the for loop
+                for (int i = 0; i < keysFromStorage.size(); i++) {
+                    JWTSigningKeyInfo currentKey = keysFromStorage.get(i);
+                    if (currentKey.algorithm.equalsIgnoreCase(algorithm.name())) {
+                        keyInfo = currentKey;
+                        break;
+                    }
+                }
+
+                // If no key was found create a new one
+                if (keyInfo == null) {
+                    keyInfo = generateAndSetKeyInStorage(noSQLStorage, algorithm);
+
+                    if (keyInfo == null) {
+                        continue;
+                    }
+                }
+
+                return keyInfo;
+            }
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    private JWTSigningKeyInfo generateKeyForAlgorithm(SupportedAlgorithms algorithm) throws NoSuchAlgorithmException,
+            UnsupportedJWTSigningAlgorithmException {
+        if (algorithm.getAlgorithmType().equalsIgnoreCase("rsa")) {
+            long currentTimeInMillis = System.currentTimeMillis();
+            Utils.PubPriKey newKey = Utils.generateNewPubPriKey();
+            return new JWTAsymmetricSigningKeyInfo(Utils.getUUID(), currentTimeInMillis, algorithm.name(), newKey.publicKey,
+                    newKey.privateKey);
+        }
+
+        throw new IllegalArgumentException();
+    }
+
+    // SQL storage
+    private JWTSigningKeyInfo generateAndSetKeyInStorage(SessionSQLStorage storage, TransactionConnection con, SupportedAlgorithms algorithm)
+            throws StorageTransactionLogicException, StorageQueryException {
+        while (true) {
+            try {
+                JWTSigningKeyInfo keyInfo = generateKeyForAlgorithm(algorithm);
+                storage.setJWTSigningKey_Transaction(con, keyInfo);
+                return keyInfo;
+            } catch (NoSuchAlgorithmException | UnsupportedJWTSigningAlgorithmException e) {
+                throw new StorageTransactionLogicException(e);
+            } catch (DuplicateKeyIdException e) {
+                // Retry with a new key id
+            }
+        }
+    }
+
+    // No SQL storage
+    private JWTSigningKeyInfo generateAndSetKeyInStorage(SessionNoSQLStorage_1 storage, SupportedAlgorithms algorithm) throws StorageTransactionLogicException, UnsupportedJWTSigningAlgorithmException, StorageQueryException {
+        while (true) {
+            try {
+                JWTSigningKeyInfo keyInfo = generateKeyForAlgorithm(algorithm);
+                boolean success = storage.setJWTSigningKeyInfo_Transaction(keyInfo);
```

Comment:
The query for the set should be something like insert keyInfo into .. if there is no entry with the matching algo (we need to do this in an atomica way)

Reply:
As discussed we will use findOneAndUpdate in the case of mongo for this